src/rgw/rgw_http_client.cc

```diff
@@ -680,28 +680,50 @@ int RGWHTTPHeadersCollector::receive_header(void * const ptr, const size_t len)
 
   if (std::string_view::npos == sep_loc) {
     /* Wrongly formatted header? Just skip it. */
+    // Also!  Normal case!  Always call for crlf at end of headers.
     return 0;
   }
 
   header_name_t name(header_line.substr(0, sep_loc));
-  if (0 == relevant_headers.count(name)) {
+  bool interesting = relevant_headers.find(name) != relevant_headers.end();
+
+  if (!interesting) {
     /* Not interested in this particular header. */
-    return 0;
-  }
 
-  const auto value_part = header_line.substr(sep_loc + 1);
+    auto hdl { header_line.length() };
+    if (hdl <= 0 || header_line[hdl-1] != '\n') {
+    } else if (--hdl > 0 && header_line[hdl-1] != '\r') {
+    } else {
+      --hdl;
```

Comment:
This logic seems unnecessarily convoluted. It also fails if header_line simply contains one newline character, as hdl, being an unsigned type, will contain a very large value (wrapping around from -1).

I think something along the lines of:
```
    auto hdl { header_line.length() };
    if (hdl > 0 && header_line[hdl-1] == '\n') {
        --hdl;
        if (hdl > 0 && header_line[hdl-1] == '\r') {
            --hdl;
        }
    }
```
would be much clearer to those following, even if you shift the first decrement into the second conditional.